test/functional/func/memoize_spec.lua

```diff
@@ -0,0 +1,120 @@
+local t = require('test.testutil')
+local n = require('test.functional.testnvim')()
+local clear = n.clear
+local exec_lua = n.exec_lua
+local eq = t.eq
+
+describe('vim.func._memoize', function()
+  before_each(clear)
+
+  it('caches function results based on their parameters', function()
+    exec_lua([[
+      _G.count = 0
+
+      local adder = vim.func._memoize('concat', function(arg1, arg2)
```

Comment:
Do we have a test that calls `vim.func._memoize('foo')` with the *same* string hash arg, and asserts that the memoized result is returned?

It's expected that `vim.func._memoize('foo')` called from anywhere in the Nvim process, from any plugin, will get the memoized result, correct? If so, that's useful for the "visibility" I mentioned earlier, because that means we have a centralized store in one place that can be inspected.

Reply:
> It's expected that `vim.func._memoize('foo')` called from anywhere in the Nvim process, from any plugin, will get the memoized result, correct? If so, that's useful for the "visibility" I mentioned earlier, because that means we have a centralized store in one place that can be inspected.

I don't think so, it's simpler than that thankfully. It would look more like this:

```lua
vim.func._memoize(tostring, function(arg) return arg end)('foo')
-- or, more realistically:
local returnme = vim.func._memoize(tostring, function(arg) return arg end)
print(returnme('foo'))
```
so you can see that the data inside the cache is local to each object and there isn't really an idea of the data being stored globally.

Just as a final note for clarity, `vim.func._memoize('foo')` is invalid, since the function expects at least two functions (and then the returned value of that will be what gets called)